cpp/dolfinx/geometry/utils.h

```diff
@@ -758,15 +758,32 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
       dolfinx::MPI::mpi_type<T>(), received_points.data(), recv_sizes.data(),
       recv_offsets.data(), dolfinx::MPI::mpi_type<T>(), forward_comm);
 
-  // Each process checks which local cell is closest and computes the
-  // squared distance to the cell
+  // Get mesh geometry for closest entity
+  const mesh::Geometry<T>& geometry = mesh.geometry();
+  if (geometry.cmaps().size() > 1)
+    throw std::runtime_error("Mixed topology not supported");
+  std::span<const T> geom_dofs = geometry.x();
+  auto x_dofmap = geometry.dofmap();
+
+  // Compute candidate cells for collisions (and extrapolation)
+  const graph::AdjacencyList<std::int32_t> candidate_collisions
+      = compute_collisions(bb, std::span<const T>(received_points.data(),
+                                                  received_points.size()));
+
+  // Each process checks which points collides with a cell on the process
   const int rank = dolfinx::MPI::rank(comm);
-  const std::vector<std::int32_t> closest_cells = compute_closest_entity(
-      bb, midpoint_tree, mesh,
-      std::span<const T>(received_points.data(), received_points.size()));
-  const std::vector<T> squared_distances = squared_distance(
-      mesh, tdim, closest_cells,
-      std::span<const T>(received_points.data(), received_points.size()));
+  std::vector<std::int32_t> cell_indicator(received_points.size() / 3);
+  std::vector<std::int32_t> closest_cells(received_points.size() / 3);
+  for (std::size_t p = 0; p < received_points.size(); p += 3)
+  {
+    std::array<T, 3> point;
+    std::copy_n(std::next(received_points.begin(), p), 3, point.begin());
+    const int colliding_cell = geometry::compute_first_colliding_cell(
```

Comment:
This code isn't easy to read. Could it be simplified?

Reply:
I've added comments to try to address this